### PR TITLE
`CidHashMap` implementation

### PR DESCRIPTION
**Pull request by jdjaustin**
## Summary of changes

Changes introduced in this pull request:

- [x] `CidHashMap` implementation with just enough methods to replace all occurrences of `HashMap<Cid, _>`.
- [x] Replace all occurrences of `HashMap<Cid, _>` with `CidHashMap<_>`.
- [x] Redefine `CidHashSet` to be `CidHashMap<Cid, ()>`.
- [x] Quickcheck properties to ensure that `CidHashMap<_>` has identical semantics to `HashMap<Cid,_>`.

## Reference issue to close (if applicable)

Closes associated task in #3202 

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

File: src/ipld/cid_hashset.rs [hunk 1]
```diff
@@ -17,23 +17,139 @@ use cid::Cid;
...
+    /// Adds a value to the set if not already present and returns whether the value was newly inserted.
     pub fn insert(&mut self, cid: Cid) -> bool {
-        match cid.try_into() {
-            Ok(CidVariant::V1DagCborBlake2b(bytes)) => self.v1_dagcbor_blake2b.insert(bytes),
-            Err(()) => self.fallback.insert(cid),
+        if self.0.contains_key(cid) {
+            false
+        } else {
+            self.0.insert(cid, ()).is_none()
         }
```
Comment (lemmih) [inline src/ipld/cid_hashset.rs:29]: This code should be rewritten. See if you can reason why. If not, ask in the slack channel.

File: src/ipld/cid_hashset.rs [hunk 2]
```diff
@@ -17,23 +17,139 @@ use cid::Cid;
...
 
+    /// Returns the number of items in the set.
     pub fn len(&self) -> usize {
-        self.v1_dagcbor_blake2b.len() + self.fallback.len()
+        self.0.len()
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct CidHashMap<V> {
```
Comment (lemmih) [inline src/ipld/cid_hashset.rs:39]: `CidHashMap` doesn't belong in the `cid_hashset` module.

File: src/ipld/cid_hashset.rs [hunk 3]
```diff
@@ -17,23 +17,139 @@ use cid::Cid;
...
+                self.v1_dagcbor_blake2b_hash_map.remove(&bytes)
+            }
+            Err(()) => self.fallback_hash_map.remove(&k),
+        }
+    }
+
+    /// Returns the number of elements the map can hold without reallocating.
+    pub fn capacity(&self) -> usize {
+        self.v1_dagcbor_blake2b_hash_map.capacity() + self.fallback_hash_map.capacity()
+    }
```
Comment (lemmih) [inline src/ipld/cid_hashset.rs:87]: This function is slightly wonky, but I'm fine with leaving it.

File: src/ipld/cid_hashset.rs [hunk 4]
```diff
@@ -17,23 +17,139 @@ use cid::Cid;
...
+
+    /// Returns the number of elements in the map.
+    pub fn len(&self) -> usize {
+        self.v1_dagcbor_blake2b_hash_map.len() + self.fallback_hash_map.len()
     }
 }
 
 #[cfg(test)]
-mod tests {}
+mod tests {
```
Comment (lemmih) [inline src/ipld/cid_hashset.rs:104]: Sets and maps usually have more than a single element. Is there a way we can test the code more broadly? Can we use arbitrary sets and maps?

**Review by lemmih — CHANGES_REQUESTED** [review]

File: src/ipld/cid_hashmap.rs [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+//   - a hash code
+//   - a length
+//   - 64 bytes pre-allocated buffer
+// Each non-buffer field takes 8 bytes with padding. So, 4*8 = 32 bytes, 32 + 64 = 96 bytes.
+//
+// However, we know that nearly all Filecoin CIDs have version=V1, codec=DAG_CBOR, code=Blake2b and
+// length=32. Taking advantage of this knowledge, we can store the vast majority of CIDs (+99.99%)
+// in one third of the usual space (32 bytes vs 96 bytes).
+#[derive(Clone, Debug, Default)]
+pub struct CidHashMap<V> {
```
Comment (aatifsyed) [inline src/ipld/cid_hashmap.rs:20]: How do you feel about implementing the standard collection ops, with due respect for size hinting when preallocating etc:
- `Extend`
- `IntoIterator`
- `FromIterator`

Comment (jdjaustin) [reply src/ipld/cid_hashmap.rs:24]: Issue #3269

File: src/ipld/cid_hashmap.rs [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+    fn contains_key(cid_vector: Vec<(Cid, u64)>, cid: Cid) {
+        let (cid_hash_map, hash_map) = generate_hash_maps(cid_vector);
+        assert_eq!(cid_hash_map.contains_key(cid), hash_map.contains_key(&cid));
+    }
+
+    #[quickcheck]
+    fn remove_key(cid_vector: Vec<(Cid, u64)>, cid: Cid) {
+        let (mut cid_hash_map, mut hash_map) = generate_hash_maps(cid_vector);
+        assert_eq!(cid_hash_map.remove(cid), hash_map.remove(&cid));
+    }
```
Comment (lemmih) [inline src/ipld/cid_hashmap.rs:139]: This test has the same problem as `contains_key`.

File: src/ipld/cid_hashmap.rs [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+    fn get_value_at_key(cid_vector: Vec<(Cid, u64)>, cid: Cid) {
+        let (cid_hash_map, hash_map) = generate_hash_maps(cid_vector);
+        assert_eq!(cid_hash_map.get(cid), hash_map.get(&cid));
+    }
+
+    #[quickcheck]
+    fn len(cid_vector: Vec<(Cid, u64)>) {
+        let (cid_hash_map, hash_map) = generate_hash_maps(cid_vector);
+        assert_eq!(cid_hash_map.len(), hash_map.len());
+    }
```
Comment (lemmih) [inline src/ipld/cid_hashmap.rs:151]: Great test!

**Review by lemmih — CHANGES_REQUESTED** [review]
When a function can return two different kinds of things (for example, `contains_key` can return either `true` or `false`), try to cover both cases with your tests. Only testing a single outcome leads to bad coverage.

File: src/ipld/cid_hashmap.rs [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+    #[quickcheck]
+    fn contains_key(cid_vector: Vec<(Cid, u64)>, cid: Cid, insert: bool) {
+        let (mut cid_hash_map, mut hash_map) = generate_hash_maps(cid_vector);
+        // Quickcheck rarely generates a key that is already present in the maps, so insert it with 50% probability to test `contains_key` with an equal distribution of results.
+        if insert {
+            cid_hash_map.insert(cid, 0);
+            hash_map.insert(cid, 0);
+        }
+        assert_eq!(cid_hash_map.contains_key(cid), hash_map.contains_key(&cid));
+    }
```
Comment (lemmih) [inline src/ipld/cid_hashmap.rs:138]: I would have created two tests: One for when `contains` returns true, and one for when it returns false. But this is okay as well.

**Review by lemmih — APPROVED** [review]

**Review by creativcoder — APPROVED** [review]

**Pull request merged**